src/Streamly/Internal/Data/Fold/Type.hs

```diff
@@ -826,9 +1018,59 @@ shortest _f1 _f2 = undefined
 -- /Unimplemented/
 --
 {-# INLINE longest #-}
-longest :: -- Monad m =>
+longest :: Monad m =>
     Fold m x a -> Fold m x a -> Fold m x a
-longest _f1 _f2 = undefined
+longest (Fold stepL beginL doneL) (Fold stepR beginR doneR) =
+    Fold step begin done
+
+    where
+
+    begin = do
+        resL <- beginL
+        resR <- beginR
+        return
+            $ case resL of
+                    Partial sl ->
+                        Partial
+                            $ case resR of
+                                Partial sr -> RunBoth sl sr
+                                Done br -> RunLeft sl br
+                    Done bl -> first (RunRight bl) resR
+
+    step (RunBoth sL sR) a = do
+        resL <- stepL sL a
+        resR <- stepR sR a
+        case resL of
+            Partial sL1 ->
+                return
+                    $ Partial
+                    $ case resR of
+                        Partial sR1 -> RunBoth sL1 sR1
+                        Done bR -> RunLeft sL1 bR
+            Done bL ->
+                return
+                    $ case resR of
+                        Partial sR1 -> Partial $ RunRight bL sR1
+                        Done bR -> Done bR
+    step (RunLeft sL bR) a = do
+        resL <- stepL sL a
+        return
+            $ case resL of
+                Partial sL1 -> Partial $ RunLeft sL1 bR
+                Done bL -> Done bL
+    step (RunRight bL sR) a = do
+        resR <- stepR sR a
+        return
+            $ case resR of
+                Partial sR1 -> Partial $ RunRight bL sR1
+                Done bR -> Done bR
```

Comment:
It would be nice if you use different constructors here that provide info.
Something like,
`LongestRunBoth`
`LongestRunOne`